blocks/bricks/bn.py

```diff
@@ -0,0 +1,350 @@
+                                        broadcastable=broadcastable)
+
+        # "beta", from the Ioffe & Szegedy manuscript.
+        self.shift = shared_floatx_nans(var_dim, name='batch_norm_shift',
+                                        broadcastable=broadcastable)
+        add_role(self.scale, BATCH_NORM_SCALE_PARAMETER)
+        add_role(self.shift, BATCH_NORM_SHIFT_PARAMETER)
+        self.parameters.append(self.scale)
+        self.parameters.append(self.shift)
+
+        # These aren't technically parameters, in that they should not be
+        # learned using the same cost function as other model parameters.
+        self.population_mean = shared_floatx_zeros(var_dim,
+                                                   name='population_mean',
+                                                   broadcastable=broadcastable)
+        self.population_stdev = shared_floatx(numpy.ones(var_dim),
+                                              name='population_stdev',
+                                              broadcastable=broadcastable)
+        add_role(self.population_mean, BATCH_NORM_POPULATION_MEAN)
+        add_role(self.population_stdev, BATCH_NORM_POPULATION_STDEV)
+
+        # Normally these would get annotated by an AnnotatingList, but they
+        # aren't in self.parameters.
+        add_annotation(self.population_mean, self)
+        add_annotation(self.population_stdev, self)
+
+    def _initialize(self):
+        self.shift_init.initialize(self.shift, self.rng)
+        self.scale_init.initialize(self.scale, self.rng)
+
+    # Needed for the Feedforward interface.
+    @property
+    def output_dim(self):
+        return self.input_dim
+
+    # The following properties allow for BatchNormalization bricks
+    # to be used directly inside of a ConvolutionalSequence.
+    @property
+    def image_size(self):
+        return self.input_dim[-2:]
+
+    @image_size.setter
+    def image_size(self, value):
+        if not isinstance(self.input_dim, collections.Sequence):
+            self.input_dim = (None,) + tuple(value)
+        else:
+            self.input_dim = (self.input_dim[0],) + tuple(value)
+
+    @property
+    def num_channels(self):
+        return self.input_dim[0]
+
+    @num_channels.setter
+    def num_channels(self, value):
+        if not isinstance(self.input_dim, collections.Sequence):
+            self.input_dim = (value,) + (None, None)
+        else:
+            self.input_dim = (value,) + self.input_dim[-2:]
+
+    def get_dim(self, name):
+        if name in ('input', 'output'):
+            return self.input_dim
+        else:
+            raise KeyError
+
+    @property
+    def num_output_channels(self):
+        return self.num_channels
+
+
+class SpatialBatchNormalization(BatchNormalization):
+    """Convenient subclass for batch normalization across spatial inputs.
+
+    Parameters
+    ----------
+    input_dim : int or tuple
```

Comment:
Also, should we put this check at the allocation stage? Lazy assignation is broken otherwise.